### PR TITLE
Add External Resource Approval API endpoint and entity

### PR DESCRIPTION
**Pull request by kotlarz**

### CODE REVIEW

File: src/main/java/com/onkibot/backend/database/entities/User.java [hunk 1]
```diff
@@ -79,4 +89,17 @@ public boolean getIsInstructor() {
...
+  }
+
+  public boolean hasApprovedExternalResource(ExternalResource externalResource) {
+    for (ExternalResource approvedExternalResource : approvedExternalResources) {
+      if (approvedExternalResource.equals(externalResource)) {
+        return true;
+      }
+    }
+    return false;
+  }
```
Comment (JamiesWhiteShirt) [inline src/main/java/com/onkibot/backend/database/entities/User.java]: getApprovedExternalResources#contains(ExternalResource)

File: src/main/java/com/onkibot/backend/database/repositories/ExternalResourceApprovalRepository.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.Collection;
+import java.util.Optional;
+import org.springframework.data.repository.CrudRepository;
+
+public interface ExternalResourceApprovalRepository
+    extends CrudRepository<ExternalResourceApproval, ExternalResourceApprovalId> {
+  Collection<ExternalResourceApproval> findByExternalResourceApprovalIdApprovalUser(
+      User approvalUser);
+
+  Optional<ExternalResourceApproval> findByExternalResourceApprovalIdExternalResource(
```
Comment (JamiesWhiteShirt) [inline src/main/java/com/onkibot/backend/database/repositories/ExternalResourceApprovalRepository.java]: Collection

File: src/main/java/com/onkibot/backend/models/ResourceModel.java [hunk 3]
```diff
@@ -20,12 +22,28 @@ public ResourceModel(Resource resource) {
...
-        resource
-            .getExternalResources()
-            .stream()
-            .map(ExternalResourceModel::new)
-            .collect(Collectors.toList());
+    this.externalResources = new ArrayList<>();
+
+    for (ExternalResource externalResource : resource.getExternalResources()) {
+      this.externalResources.add(new ExternalResourceModel(externalResource, false));
+    }
```
Comment (JamiesWhiteShirt) [inline src/main/java/com/onkibot/backend/models/ResourceModel.java]: I want streams

Comment (JamiesWhiteShirt) [reply src/main/java/com/onkibot/backend/models/ResourceModel.java]: pls

File: src/main/java/com/onkibot/backend/models/ResourceModel.java [hunk 4]
```diff
@@ -20,12 +22,28 @@ public ResourceModel(Resource resource) {
...
+    this.body = resource.getBody();
+    this.publisherUser = new UserModel(resource.getPublisherUser());
+    this.externalResources = new ArrayList<>();
+
+    for (ExternalResource externalResource : resource.getExternalResources()) {
+      if (sessionUser.hasApprovedExternalResource(externalResource)) {
+        this.externalResources.add(new ExternalResourceModel(externalResource, true));
+      } else {
+        this.externalResources.add(new ExternalResourceModel(externalResource, false));
+      }
```
Comment (JamiesWhiteShirt) [inline src/main/java/com/onkibot/backend/models/ResourceModel.java]: Ditto

File: src/main/java/com/onkibot/backend/api/ExternalResourceController.java [hunk 5]
```diff
@@ -33,22 +37,30 @@ ExternalResourceModel get(
...
-        .map(ExternalResourceModel::new)
-        .collect(Collectors.toList());
+    List<ExternalResource> externalResources = resource.getExternalResources();
+    List<ExternalResourceModel> externalResourceModels = new ArrayList<>();
+    for (ExternalResource externalResource : externalResources) {
+      externalResourceModels.add(
+          new ExternalResourceModel(
+              externalResource, hasApprovedExternalResource(externalResource, session)));
+    }
+    return externalResourceModels;
```
Comment (JamiesWhiteShirt) [inline src/main/java/com/onkibot/backend/api/ExternalResourceController.java]: STREAMS

**Review by JamiesWhiteShirt — CHANGES_REQUESTED** [review]

**Comment by kotlarz** [comment]
@JamiesWhiteShirt The reason it has been marked as failed is because the coverage went down as there are no tests for the implemented code.

**Pull request merged**